runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/status/WorkerStatusClient.java

```diff
@@ -0,0 +1,146 @@
+ * limitations under the License.
+ */
+package org.apache.beam.runners.fnexecution.status;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.WorkerStatusRequest;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.WorkerStatusResponse;
+import org.apache.beam.sdk.fn.stream.SynchronizedStreamObserver;
+import org.apache.beam.vendor.grpc.v1p21p0.io.grpc.stub.StreamObserver;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Client for handling requests and responses over Fn Worker Status Api between runner and SDK
+ * Harness.
+ */
+public class WorkerStatusClient implements Closeable {
+
+  public static final Logger LOG = LoggerFactory.getLogger(WorkerStatusClient.class);
+  private final StreamObserver<WorkerStatusRequest> requestReceiver;
+  private final Map<String, CompletableFuture<WorkerStatusResponse>> responseQueue =
+      new ConcurrentHashMap<>();
+  private final String workerId;
+  private AtomicBoolean isClosed = new AtomicBoolean(false);
+
+  private WorkerStatusClient(String workerId, StreamObserver<WorkerStatusRequest> requestReceiver) {
+    this.requestReceiver = SynchronizedStreamObserver.wrapping(requestReceiver);
+    this.workerId = workerId;
+  }
+
+  /**
+   * Create new status api client with SDK Harness worker id and request observer.
+   *
+   * @param workerId SDK Harness worker id.
+   * @param requestObserver The outbound request observer this client uses to send new status
+   *     requests to its corresponding SDK Harness.
+   * @return {@link WorkerStatusClient}
+   */
+  public static WorkerStatusClient forRequestObserver(
+      String workerId, StreamObserver<WorkerStatusRequest> requestObserver) {
+    return new WorkerStatusClient(workerId, requestObserver);
+  }
+
+  /**
+   * Get the latest sdk worker status from the client's corresponding SDK Harness. A random UUID
+   * will be used to specify the request_id field.
+   *
+   * @return {@link CompletableFuture} of the SDK Harness status response.
+   */
+  public CompletableFuture<WorkerStatusResponse> getWorkerStatus() {
+    WorkerStatusRequest request =
+        WorkerStatusRequest.newBuilder().setId(UUID.randomUUID().toString()).build();
+    return getWorkerStatus(request);
+  }
+
+  /**
+   * Get the latest sdk worker status from the client's corresponding SDK Harness with request.
+   *
+   * @param request WorkerStatusRequest to be sent to SDK Harness.
+   * @return {@link CompletableFuture} of the SDK Harness status response.
+   */
+  @VisibleForTesting
+  public CompletableFuture<WorkerStatusResponse> getWorkerStatus(WorkerStatusRequest request) {
```

Comment:
we can keep visibility to default to allow visibility in test.

Reply:
BeamWorkerStatusGrpcService.java was in another package and this public function provides more flexibility in its test which probably not easy to get with default access.

Reply:
SG